libs/auth/src/cast_vote_record_hashes.ts

```diff
@@ -0,0 +1,308 @@
+import { Buffer } from 'buffer';
+import fs from 'fs/promises';
+import { sha256 } from 'js-sha256';
+import path from 'path';
+import { groupBy } from '@votingworks/basics';
+import { Client } from '@votingworks/db';
+
+/**
+ * A representation of a file
+ */
+export interface File {
+  fileName: string;
+  fileContents: string | Buffer;
+}
+
+/**
+ * Computes a hash for a single cast vote record, incorporating the cast vote record's directory
+ * name, file names, and file contents. The directory summary that we hash specifically mirrors the
+ * output of:
+ *
+ * find <directoryName> -type f | sort | xargs sha256sum
+ */
+export function computeSingleCastVoteRecordHash({
+  directoryName,
+  files,
+}: {
+  directoryName: string;
+  files: File[];
+}): string {
+  const filesSorted = [...files].sort((file1, file2) =>
+    file1.fileName.localeCompare(file2.fileName)
+  );
+  const fileHashes: string[] = [];
+  for (const { fileName, fileContents } of filesSorted) {
+    fileHashes.push(
+      `${sha256(fileContents)}  ${path.join(directoryName, fileName)}\n`
+    );
+  }
+  const directorySummary = fileHashes.join('');
+  return sha256(directorySummary);
+}
+
+/**
+ * The input to {@link computeCombinedHash}
+ */
+export type HashesToCombine = Array<{ hash: string; sortKey: string }>;
+
+/**
+ * Sorts the provided hashes using the specified sort key, concatenates the hashes, and hashes the
+ * result, yielding a combined hash
+ */
+export function computeCombinedHash(hashesToCombine: HashesToCombine): string {
+  return sha256(
+    [...hashesToCombine]
+      .sort((entry1, entry2) => entry1.sortKey.localeCompare(entry2.sortKey))
```

Comment:
Framing this another way, this sort is equivalent to the `order by` in our DB queries and the sort of our file names:

- https://github.com/votingworks/vxsuite/blob/338374416a9023ed7e16934f4f51a3406b1ef0ed/libs/auth/src/cast_vote_record_hashes.ts#L140
- https://github.com/votingworks/vxsuite/blob/338374416a9023ed7e16934f4f51a3406b1ef0ed/libs/auth/src/cast_vote_record_hashes.ts#L29-L31

Reply:
ah, if you're just sorting the filenames pre-hashing, then great.

My read of the code is that this would also sort the hashes that are combined at level 2 by hash... am I wrong? If so, great.

Reply:
Yup, no sorting by hash at any point - but the code is dense/abstract enough at points that I understand where that confusion comes from! Planning to push some small tweaks to make things clearer